pulp_rpm/app/tasks/synchronizing.py

```diff
@@ -492,29 +589,47 @@ def _handle_distribution_tree(declarative_content):
-            if relations_exist:
-                # existing content which was retrieved from the db at earlier stages
-                continue
-
-            future_relations = declarative_content.extra_data
-            update_collections = future_relations.get('collections') or {}
-            update_references = future_relations.get('references') or []
-
-            for update_collection, packages in update_collections.items():
-                update_collection.update_record = update_record
-                update_collections_to_save.append(update_collection)
-                for update_collection_package in packages:
-                    update_collection_package.update_collection = update_collection
-                    update_collection_packages_to_save.append(update_collection_package)
-
-            for update_reference in update_references:
-                update_reference.update_record = update_record
-                update_references_to_save.append(update_reference)
+            elif isinstance(declarative_content.content, UpdateRecord):
+                update_record = declarative_content.content
+
+                relations_exist = (
+                    update_record.collections.count() or update_record.references.count()
+                )
+                if relations_exist:
+                    # existing content which was retrieved from the db at earlier stages
+                    continue
+
+                future_relations = declarative_content.extra_data
+                update_collections = future_relations.get('collections') or {}
+                update_references = future_relations.get('references') or []
+
+                for update_collection, packages in update_collections.items():
+                    update_collection.update_record = update_record
+                    update_collections_to_save.append(update_collection)
+                    for update_collection_package in packages:
+                        update_collection_package.update_collection = update_collection
+                        update_collection_packages_to_save.append(update_collection_package)
+
+                for update_reference in update_references:
+                    update_reference.update_record = update_record
+                    update_references_to_save.append(update_reference)
+
+            elif isinstance(declarative_content.content, Modulemd):
+                if 'package_relation' in declarative_content.extra_data:
+                    for pkg in declarative_content.extra_data['package_relation']:
+                        try:
+                            declarative_content.content.packages.add(pkg.content)
+                        except (ValueError, IntegrityError):
+                            pass
+
+            elif isinstance(declarative_content.content, Package):
+                if 'modulemd_relation' in declarative_content.extra_data:
+                    declarative_content.content.is_modular = True
```

Comment:
I think this should be done earlier, in the first stage, I left the comment there.  We are inthe post_save hook here, so object is already saved, so the changes to attributes of declarative_content.content won't persist.

Reply:
refactored to be done at first stage when Package is checked against modules if it is part of module